### PR TITLE
FEATURE: Unique client id for telemetry events

### PR DESCRIPTION
**Pull request by vikramdevtron**
# Description

Get api for fetching client id, for dashboard to send event to posthog.

Fixes # (issue): https://github.com/devtron-labs/devtron/issues/530

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Get client Id

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have tested it for all user roles

### CODE REVIEW

File: client/telemetry/TelemetryEventClient.go [hunk 1]
```diff
@@ -283,3 +284,53 @@ func (impl *TelemetryEventClientImpl) HeartbeatEventForTelemetry() {
...
+		}
+	}
+
+	if cm == nil {
+		impl.logger.Errorw("configmap not found while getting unique client id", "cm", cm)
+		return nil, err
+	}
+	dataMap := cm.Data
+	ucid := dataMap[DevtronUniqueClientIdConfigMapKey]
+	data := &PosthogData{
```
Comment (nishant-d) [inline client/telemetry/TelemetryEventClient.go:326]: cache UIID, fetching data from cm for each request will put unnecessary load on api server

File: client/telemetry/TelemetryEventClient.go [hunk 2]
```diff
@@ -283,3 +284,53 @@ func (impl *TelemetryEventClientImpl) HeartbeatEventForTelemetry() {
 		Properties: prop,
 	})
 }
+
+
+func (impl *TelemetryEventClientImpl) GetUCID() (*PosthogData, error) {
+	clusterBean, err := impl.clusterService.FindOne(cluster.ClusterName)
```
Comment (nishant-d) [inline client/telemetry/TelemetryEventClient.go:290]: we should write a single method for fetching `UCID`
duplicate code

File: api/restHandler/TelemetryRestHandler.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	userAuthService      user.UserAuthService
+	validator            *validator.Validate
+	logger               *zap.SugaredLogger
+	enforcer             rbac.Enforcer
+	natsClient           *pubsub.PubSubClient
+	userService          user.UserService
+	telemetryEventClient telemetry.TelemetryEventClient
+}
+
+func NewTelemetryRestHandlerImpl(userAuthService user.UserAuthService, validator *validator.Validate,
```
Comment (nishant-d) [inline api/restHandler/TelemetryRestHandler.go:44]: do we need all these arguments ?

File: api/router/TelemetryRouter.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+func NewTelemetryRouterImpl(logger *zap.SugaredLogger, handler restHandler.TelemetryRestHandler) *TelemetryRouterImpl {
+	router := &TelemetryRouterImpl{
+		handler: handler,
+	}
+	return router
+}
+
+func (router TelemetryRouterImpl) initTelemetryRouter(telemetryRouter *mux.Router) {
+	telemetryRouter.Path("/ucid").
```
Comment (nishant-d) [inline api/router/TelemetryRouter.go:43]: do we require authentication for this URL?

**Review by nishant-d — CHANGES_REQUESTED** [review]

File: api/restHandler/TelemetryRestHandler.go (hunk 3 above)
Comment (vikramdevtron) [reply api/restHandler/TelemetryRestHandler.go:44]: removed unused

File: api/router/TelemetryRouter.go (hunk 4 above)
Comment (vikramdevtron) [reply api/router/TelemetryRouter.go:43]: user login enough for authentication for this.

File: client/telemetry/TelemetryEventClient.go (hunk 2 above)
Comment (vikramdevtron) [reply client/telemetry/TelemetryEventClient.go:290]: fixed

File: client/telemetry/TelemetryEventClient.go (hunk 1 above)
Comment (vikramdevtron) [reply client/telemetry/TelemetryEventClient.go:255]: implemented

**Pull request merged**